nexus/db-queries/src/db/datastore/snapshot_replacement.rs

```diff
@@ -0,0 +1,1428 @@
+                                record.replacement_state,
+                                record.operating_saga_id,
+                            ))))
+                        }
+                    }
+                }
+            })
+            .await
+            .map_err(|e| match e {
+                TxnError::CustomError(error) => error,
+
+                TxnError::Database(error) => {
+                    public_error_from_diesel(error, ErrorHandler::Server)
+                }
+            })
+    }
+
+    pub async fn create_snapshot_replacement_step(
+        &self,
+        opctx: &OpContext,
+        request_id: Uuid,
+        volume_id: Uuid,
+    ) -> Result<Uuid, Error> {
+        let request = SnapshotReplacementStep::new(request_id, volume_id);
+        let request_id = request.id;
+
+        self.insert_snapshot_replacement_step(opctx, request).await?;
+
+        Ok(request_id)
+    }
+
+    pub async fn insert_snapshot_replacement_step(
+        &self,
+        opctx: &OpContext,
+        request: SnapshotReplacementStep,
+    ) -> Result<(), Error> {
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        let err = OptionalError::new();
+        self.transaction_retry_wrapper("insert_snapshot_replacement_step")
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                let request = request.clone();
+
+                async move {
+                    use db::schema::snapshot_replacement_step::dsl;
+                    use db::schema::volume_repair::dsl as volume_repair_dsl;
+
+                    // Skip inserting this new record if we found another
+                    // snapshot replacement step with this volume in the step's
+                    // `old_snapshot_volume_id`, as that means we're duplicating
+                    // the replacement work: that volume will be garbage
+                    // collected later. There's a unique index that will prevent
+                    // the same step being inserted with the same volume id.
+
+                    let maybe_record = dsl::snapshot_replacement_step
+                        .filter(dsl::old_snapshot_volume_id.eq(request.volume_id))
+                        .get_result_async::<SnapshotReplacementStep>(
+                            &conn,
+                        )
+                        .await
+                        .optional()?;
+
+                    if let Some(found_record) = maybe_record {
+                        return Err(err.bail(Error::conflict(format!(
+                            "{:?} already referenced in old snapshot volume for \
+                            request {:?}",
+                            request.volume_id,
+                            found_record.id,
+                        ))));
+                    }
+
+                    // The snapshot replacement step saga could invoke a volume
+                    // replacement: create an associated volume repair record.
```

Comment:
`could`?
Is this part of the paranoia changes?

Reply:
`could` here meaning if there isn't an associated running Upstairs in a Propolis we don't need to invoke any replacement.